pkg/plugins/discovery/k8s/controllers/pod_converter.go

```diff
@@ -130,10 +126,107 @@ func InboundInterfacesFor(pod *kube_core.Pod, services []*kube_core.Service, isG
+			outbounds = append(outbounds, &mesh_proto.Dataplane_Networking_Outbound{
+				Address: service.Spec.ClusterIP,
+				Port:    port,
+				Service: serviceTag,
+			})
+		}
+	}
+
+	directAccessOutbounds := directAccessOutbounds(directAccessServices, endpoints)
+	return append(outbounds, directAccessOutbounds...), nil
+}
+
+func directAccessServices(pod *kube_core.Pod) map[string]bool {
+	result := map[string]bool{}
+	servicesRaw := pod.GetAnnotations()[injector_metadata.KumaDirectAccess]
+	services := strings.Split(servicesRaw, ",")
+	for _, service := range services {
+		result[service] = true
+	}
+	return result
+}
+
+// Generate outbound listeners for every endpoint of services.
+// This will enable consuming applications via transparent proxy by PodIP instead of ClusterIP of its service
+// Generating listener for every endpoint will cause XDS snapshot to be huge therefore it should be used only if really needed
+func directAccessOutbounds(services map[string]bool, endpointsByService EndpointsByService) []*mesh_proto.Dataplane_Networking_Outbound {
+	var sortedServices []string // service should be sorted so we generate consistent every time
+	if services[injector_metadata.KumaDirectAccessAll] {
+		sortedServices = endpointsByService.Services()
+	} else {
+		sortedServices = stringSetToSortedList(services)
+	}
+
+	var outbounds []*mesh_proto.Dataplane_Networking_Outbound
+	for _, service := range sortedServices {
+		// services that are not found will be ignored
+		for _, endpoint := range endpointsByService[service] {
+			outbounds = append(outbounds, &mesh_proto.Dataplane_Networking_Outbound{
+				Address: endpoint.Address,
+				Port:    endpoint.Port,
+				Service: service,
+			})
+		}
+	}
+	return outbounds
+}
+
+func isHeadlessService(svc *kube_core.Service) bool {
+	return svc.Spec.ClusterIP == "None"
+}
+
+func (p *PodConverter) k8sService(serviceTag string) (*kube_core.Service, uint32, error) {
+	host, port, err := mesh_proto.ServiceTagValue(serviceTag).HostAndPort()
+	if err != nil {
+		return nil, 0, errors.Errorf("failed to parse `service` tag %q", serviceTag)
+	}
+	name, ns, err := ParseServiceFQDN(host)
+	if err != nil {
+		return nil, 0, errors.Errorf("failed to parse `service` host %q as FQDN", host)
+	}
+
+	svc := &kube_core.Service{}
+	svcKey := kube_client.ObjectKey{Namespace: ns, Name: name}
+	if err := p.ServiceGetter.Get(context.Background(), svcKey, svc); err != nil {
+		return nil, 0, errors.Wrapf(err, "failed to get Service %q", svcKey)
+	}
+	return svc, port, nil
+}
+
+type Endpoint struct {
+	Address string
+	Port    uint32
+}
+
+type EndpointsByService map[string][]Endpoint
+
+func (e EndpointsByService) Services() []string {
+	list := make([]string, 0, len(e))
+	for key := range e {
+		list = append(list, key)
+	}
+	sort.Strings(list)
+	return list
+}
 
-	allServiceTags := make(map[string]bool)
+func endpointsByService(others []*mesh_k8s.Dataplane) EndpointsByService {
```

Comment:
From the perspective of that method, this parameter should be called `dataplanes` or `dps`. It doesn't know about "others"

Reply:
make sense, done